src/kms.rs

```diff
@@ -0,0 +1,329 @@
+                &[self.connector],
+                self.crtc.mode(),
+            )
+            .swbuf_err("failed to set CRTC")?;
+
+        self.buffer = Some(buffer);
+
+        Ok(())
+    }
+
+    /// Fetch the buffer from the window.
+    pub(crate) fn fetch(&mut self) -> Result<Vec<u32>, SoftBufferError> {
+        // TODO: Implement this!
+        Err(SoftBufferError::Unimplemented)
+    }
+
+    /// Get a mutable reference to the buffer.
+    pub(crate) fn buffer_mut(&mut self) -> Result<BufferImpl<'_>, SoftBufferError> {
+        // Map the dumb buffer.
+        let set = self
+            .buffer
+            .as_mut()
+            .expect("Must set size of surface before calling `buffer_mut()`");
+        let size = set.size();
+        let mapping = self
+            .display
+            .map_dumb_buffer(&mut set.db)
+            .swbuf_err("failed to map dumb buffer")?;
+
+        Ok(BufferImpl {
+            mapping,
+            size,
+            fb: set.fb,
+            display: &self.display,
+            zeroes: &set.zeroes,
+        })
+    }
+}
+
+impl Drop for KmsImpl {
+    fn drop(&mut self) {
+        // Map the CRTC to the information that was there before.
+        self.display
+            .set_crtc(
+                self.crtc.handle(),
+                self.crtc.framebuffer(),
+                self.crtc.position(),
+                &[self.connector],
+                self.crtc.mode(),
+            )
+            .ok();
+    }
+}
+
+impl BufferImpl<'_> {
+    #[inline]
+    pub fn pixels(&self) -> &[u32] {
+        // drm-rs doesn't let us have the immutable reference... so just use a bunch of zeroes.
+        // TODO: There has to be a better way of doing this!
+        self.zeroes
+    }
+
+    #[inline]
+    pub fn pixels_mut(&mut self) -> &mut [u32] {
+        bytemuck::cast_slice_mut(self.mapping.as_mut())
+    }
+
+    #[inline]
+    pub fn age(&self) -> u8 {
+        todo!()
+    }
+
+    #[inline]
+    pub fn present_with_damage(self, damage: &[crate::Rect]) -> Result<(), SoftBufferError> {
+        let rectangles = damage
+            .iter()
+            .map(|&rect| {
+                let err = || SoftBufferError::DamageOutOfRange { rect };
+                Ok(drm_sys::drm_clip_rect {
+                    x1: rect.x.try_into().map_err(|_| err())?,
+                    y1: rect.y.try_into().map_err(|_| err())?,
+                    x2: rect
+                        .x
+                        .checked_add(rect.width.get())
+                        .and_then(|x| x.try_into().ok())
+                        .ok_or_else(err)?,
+                    y2: rect
+                        .y
+                        .checked_add(rect.height.get())
+                        .and_then(|y| y.try_into().ok())
+                        .ok_or_else(err)?,
+                })
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        self.display
+            .dirty_framebuffer(self.fb, &rectangles)
```

Comment:
I think this works but gives tearing. I suggest using two dumb buffers and flip between them. You could say that `present()` in soft buffer just posts it, but it's up to the application to wait for the page flip (although that's also tricky).

Reply:
I think you just allocate two dumb buffers instead of one,map the one that you want to render into, create a frame buffer, and set it via set_crtc() when you're ready. In the next round you render into the second buffer, create a frame buffer, and then make it the new front buffer by calling `page_flip()`. 

As if you're using gbm, except that you're doing managing a back and front buffer yourself.

I don't know if the code we're using is 100% correct, but with gbm seems to work well like this: https://github.com/slint-ui/slint/blob/f8e9d6237999910db905b4ba84d2cd3520268d44/internal/backends/linuxkms/display/egldisplay.rs#L51

Without gbm I think you'd just allocate two dumb buffers (like you currently do in new()) and otherwise `present()` is pretty much the same. Perhaps some more magic in gbm's lock_buffer(), but I suspect that's more GL related (flushing pipeline?) than buffer related.

Take this with a grain of salt though ;-). I've been meaning to try to do this myself with Rust code ("double dumb buffers" with software rendering), but haven't gotten around it yet.